pandas/core/indexes/base.py

```diff
@@ -1077,7 +1077,11 @@ def astype(self, dtype, copy: bool = True):
             with rewrite_exception(type(values).__name__, type(self).__name__):
                 new_values = values.astype(dtype, copy=copy)
 
-        elif is_float_dtype(self.dtype) and needs_i8_conversion(dtype):
+        elif (
+            is_float_dtype(self.dtype)
+            and needs_i8_conversion(dtype)
+            and not np.isnan(values).any()
```

Comment:
can you add a comment about why this check is needed

Reply:
also use self.hasnans as it may be cached

Reply:
> can you add a comment about why this check is needed

Hi, I just notice this. it is my mistake to add this.